rfcs/text/0000-basic-typed-contracts.md

```diff
@@ -0,0 +1,249 @@
+*output type*. For example, the input type of `exact-integer?` is
+`Any` and its output type is `Integer`.
+
+Back to type checking the combined contract, `(and/c exact-integer?
+even?)`. From the types above, we know that what comes out of
+`exact-integer?` can safely go in to `even?`. And since everything
+lines up, we take the the input type of the first contract type
+(`Integer`) and the output type of the last contract (`Integer`) and
+give the whole contract type `(Contract Any Integer)`.
+
+## Contracts on higher-order types
+
+If a function only consumes and produces positive reals, a Racket
+programmer might express that information using the contract `(->/c
+positive? positive?)`. In Typed Racket, that same information could be
+expressed with the type `(-> Positive-Real Positive-Real)`. So
+although that exact contract might not make sense in a Typed Racket
+program, we still want to give it a type---the path from Racket to
+Typed Racket isn't always smooth so we might be able to help the
+Racket programmer if we give a meaningful type to that contract.
+
+Starting with the function `positive?`, its type is `(-> Real
+Boolean)` and it has a positive proposition for `Positive-Real`. So
+based on what we saw in the previous section, it has the contract type
+`(Contract Real Positive-Real)`.
+
+The function contract combinator `->/c` wraps whatever function it is
+applied to, ensuring that a caller provides arguments that satisfy the
+domain contract and that the function returns a result that satisfies
+the range contract; Values flow from the caller through the domain
+contract into the function, and out of the function and through the
+range contract.
+
+This value flow is reflected in the type of a function contract. For
+our `(->/c positive? positive?)` the type is
+
+    (Contract (-> {2} Positive-Real Real {3})
+              (-> {1} Real Positive-Real {4}))
+
+In the output type of the contract, here a function type, the domain
+type is `Real` (marked `{1}`). This is because after applying the
+contract to a function, calling the result with a value will apply the
+domain contract (with input type `Real`) to that value. If the value
+passes the domain contract, we know it is a `Positive-Real` by the
+domain contract's output type, which is reflected in the input type's
+domain contract (marked `{2}`). This means that a function we apply
+this contract to has to accept `Positive-Real` values.
+
+In the input type of the contract, the range of the type is `Real`
+(marked `{3}`) because applying the contract to a function is only
+allowed if that function returns a value that can flow through the
+range contract (whose input type is `Real`). Finally, in the output
+type of the contract the range type is `Positive-Real` (marked `{4}`)
+because that's the output type of the range contract. This means that
+if we apply this contract to a function, then the resuling function
+returns at least a `Positive-Real` if it returns at all.
+
```

Comment:
Add quick examples for lists, structs, and/or vectors?

I'm guessing: functions are a great example because a function type has input & output. Other contract-types should be simpler, but it'd be nice to have that confirmed.

Reply:
I'll add a list example.

As far as structs go: No support yet for `struct/c`, and I might need to provide a variant of `contract-out` that does some additional checks; At the moment the PR just reprovides Racket's `contract-out`. I updated the RFC with notes about that.

For vectors, are you interested in mutable vectors?

Reply:
Mutable specifically